validator/index.js

```diff
@@ -25,50 +25,233 @@ const path = require('path');
-  };
-  if (fileOrUrl === '-') {
-    digestChunksFrom(process.stdin);
-    process.stdin.resume();
-    return;
-  }
-  if (fileOrUrl.startsWith('http://') || fileOrUrl.startsWith('https://')) {
-    const clientModule = fileOrUrl.startsWith('http://') ? http : https;
-    const req = clientModule.get(fileOrUrl, (response) => {
-      if (response.statusCode !== 200) {
-        onFailure('HTTP ' + response.statusCode);
-        return;
+    readable.setEncoding('utf8');
+    readable.on('data', (chunk) => { chunks.push(chunk); });
+    readable.on('end', () => { resolve(chunks.join('')); });
+    readable.on('error', (error) => { reject(error); });
+  });
+}
+
+/**
+ * Creates a promise which reads from standard input.
+ * @returns {!Promise<!string|Error>}
+ */
+function readFromStdin() {
+  return new Promise(function(resolve, reject) {
+    readFromReadable(process.stdin)
+        .then((data) => {
+          process.stdin.resume();
+          resolve(data);
+        })
+        .catch(reject);
+  });
+}
+
+/**
+ * Creates a promise which reads from a URL or more precisely, fetches
+ * the contents located at the URL by using the 'http' or 'https' module.
+ * Any HTTP status other than 200 is interpreted as an error.
+ * @param {!string} url
+ * @returns {!Promise<!string|Error>}
+ */
+function readFromUrl(url) {
+  return new Promise((resolve, reject) => {
+    const clientModule = url.startsWith('http://') ? http : https;
+    clientModule.get(url, (response) => {
+      if (response.statusCode != 200) {
+        response.resume();
+        reject(new Error('HTTP Status ' + response.statusCode));
+      } else {
+        readFromReadable(response).then(resolve);
       }
-      digestChunksFrom(response);
     });
-    req.on('error', (e) => { onFailure(e.message); });
-    return;
+  });
+}
+
+/**
+ * ValidationResult is the record computed by the validator for each
+ * document. It contains an overall status (PASS/FAIL) and the list
+ * of errors, if any.
+ * @export
+ */
+class ValidationResult {
+  constructor() {
+    /**
+     * Possible values are 'UNKNOWN', 'PASS', and 'FAIL'.
+     * @type {!string}
+     */
+    this.status = 'UNKNOWN';
+    /** @type {!Array<!ValidationError>} */
+    this.errors = [];
   }
-  fs.readFile(fileOrUrl, 'utf8', (err, data) => {
-    if (err) {
-      onFailure(err.message);
-    } else {
-      onSuccess(data);
+}
+
+/**
+ * Each validation error describes a specific problem in a validated
+ * document.
+ * @export
+ */
+class ValidationError {
+  constructor() {
+    /**
+     * The severity of the error - possible values are 'UNKNOWN_SEVERITY',
+     * 'ERROR', and 'WARNING'.
+     */
+    this.severity = 'UNKNOWN_SEVERITY';
+    /**
+     * The line number at which the error was seen (1 is the first line).
+     */
+    this.line = 1;
+    /**
+     * The column number at which the error was seen (0 is the first column).
+     */
+    this.col = 0;
```

Comment:
I came across this before too. The line numbers are one-based and column numbers are zero-based. This is fairly typical for javascript errors.